drv/gimlet-seq-server/build.rs

```diff
@@ -12,9 +14,148 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
 
     let out = &PathBuf::from(env::var_os("OUT_DIR").unwrap());
     fs::write(out.join("fpga.bin.rle"), compressed)?;
+
+    let disposition = build_i2c::Disposition::Devices;
+
+    if let Err(e) = build_i2c::codegen(disposition) {
+        println!("code generation failed: {}", e);
+        std::process::exit(1);
+    }
+
+    fs::write(out.join("gimlet_regs.rs"), regs()?)?;
+    println!("cargo:rerun-if-changed=fpga.bin");
+
+    idol::server::build_server_support(
+        "../../idl/gimlet-seq.idol",
+        "server_stub.rs",
+        idol::server::ServerStyle::InOrder,
+    )?;
+
     Ok(())
 }
 
+#[derive(Debug, Deserialize)]
+#[serde(tag = "type", rename_all = "lowercase")]
+enum Node {
+    Addrmap {
+        children: Vec<Node>,
+    },
+    Reg {
+        inst_name: String,
+        addr_offset: usize,
+        regwidth: usize,
+        children: Vec<Node>,
+    },
+    Field {
+        inst_name: String,
+        lsb: usize,
+        msb: usize,
+    },
+}
+
+fn regs() -> Result<String, Box<dyn std::error::Error>> {
+    let mut output = String::new();
+    let regs = include_str!("gimlet_regs.json");
+
+    let node: Node = serde_json::from_str(regs)?;
+
+    let children = if let Node::Addrmap { children } = node {
+        children
+    } else {
+        panic!("top-level node is not addrmap");
+    };
+
+    writeln!(
+        &mut output,
+        r##"
+#[allow(non_camel_case_types)]
+pub enum Addr {{"##
+    )?;
+
+    for child in children.iter() {
+        if let Node::Reg {
+            inst_name,
+            addr_offset,
+            ..
+        } = child
+        {
+            writeln!(&mut output, "    {} = 0x{:x},", inst_name, addr_offset)?;
+        } else {
+            panic!("unexpected child {:?}", child);
+        }
+    }
+
+    println!("cargo:rerun-if-changed=gimlet_regs.json");
```

Comment:
This is not necessary, `include_str!` does it as a side effect.

Reply:
TIL!